snapshots/lcow/lcow.go

```diff
@@ -311,7 +314,14 @@ func (s *snapshotter) createSnapshot(ctx context.Context, kind snapshots.Kind, k
 	if err != nil {
 		return nil, err
 	}
-	defer t.Rollback()
+
+	defer func() {
+		if err != nil && t != nil {
+			if rerr := t.Rollback(); rerr != nil {
+				log.G(ctx).WithError(rerr).Warn("failed to rollback transaction")
+			}
+		}
+	}()
```

Comment:
You aren't using named return values here so the err != nil check in here would only work for the last err set in the function.

Reply:
We return if the err isn't nil from `TransactionContext` though, the defer wouldn't get run as it doesn't exist yet. And then any error following it should work fine with this afaik. I'll just change it back to defer t.Rollback anyways, it wasn't hurting anyone 😆 Just saw this used in the overlay snapshotter

Reply:
Not sure I follow your point Danny. Regarding what Kathryn is saying, it's pretty much always a good idea to use named return values if you have a `defer` that tries to check if an error is being returned from the function. Otherwise, it's really easy to fall victim to shadowing.

Consider this:

```
func foo() error {
	var err error
	defer func() {
		if err != nil {
			fmt.Printf("defer saw an error: %s\n", err)
		}
	}()
	if err := errors.New("generic error"); err != nil {
		return err
	}
	return nil
}

func main() {
	if err := foo(); err != nil {
		fmt.Printf("main saw an error: %s\n", err)
	}
}
```

Output:
```
> go run .
main saw an error: generic error
```

Because the `err` value is assigned in an `if` conditional, it shadows the top-level `err` declaration, so the `defer` statement doesn't see it.

Reply:
Donezo!

Reply:
No I understand what was being said now, I just reverted to what it was before as the Rollback doesn't do anything if it's been committed. I added the named return value anyways as I also added a cleanup of the snapshot directory if we fail. Thanks guys!